### PR TITLE
Add str.__rmod__() method. #190

### PR DESCRIPTION
**Pull request by ChJR**
CPython returns "NotImplemented" string when arguments supplied.

### CODE REVIEW

File: tests/snippets/strings.py [hunk 1]
```diff
@@ -293,3 +293,12 @@ def try_mutate_str():
...
 assert_raises(StopIteration, lambda: next(str_iter_reversed))
+
+assert str.__rmod__('%i', 30) == "NotImplemented"
+
+typeError = False
+try:
+    str.__rmod__(30, '%i')
+except TypeError:
+    typeError = True
+assert typeError
```
Comment (coolreader18) [inline tests/snippets/strings.py]: You can use `assertRaises` from `testutils` for this:

```python
from testutils import assertRaises

...

with assertRaises(TypeError):
    ...
```

File: vm/src/obj/objstr.rs [hunk 2]
```diff
@@ -518,6 +518,18 @@ impl PyString {
...
+    #[pymethod(name = "__rmod__")]
+    fn rmod(&self, values: PyObjectRef, vm: &VirtualMachine) -> PyResult {
+        let format_string_text = &self.value;
+        let _format_string = CFormatString::from_str(format_string_text)
+            .map_err(|err| vm.new_value_error(format!("{}", err)))?;
+
+        match do_cformat(vm, _format_string, values.clone()) {
+            Ok(_format_string) => Ok(vm.new_str("NotImplemented".to_string())),
+            Err(err) => Err(err)
+        }
```
Comment (coolreader18) [inline vm/src/obj/objstr.rs]: I think you misunderstood what you were seeing in the python shell -- in CPython, `str.__rmod__(str, int)` doesn't return `"NotImplemented"`, it returns the actual value `NotImplemented`, which you can access from Rust as `vm.ctx.not_implemented()`.

Comment (ChJR) [reply vm/src/obj/objstr.rs]: I misunderstood that. I will fix and re-open this pull request.

File: tests/snippets/strings.py (hunk 1 above)
Comment (ChJR) [reply tests/snippets/strings.py]: Thank you for sharing. I will use your proposal for this PR.

**Comment by ChJR** [comment]
I improved PR code quality.
Thanks to @coolreader18

File: vm/src/obj/objstr.rs [hunk 3]
```diff
@@ -518,6 +518,18 @@ impl PyString {
         do_cformat(vm, format_string, values.clone())
     }
 
+    #[pymethod(name = "__rmod__")]
+    fn rmod(&self, values: PyObjectRef, vm: &VirtualMachine) -> PyResult {
+        let format_string_text = &self.value;
+        let _format_string = CFormatString::from_str(format_string_text)
+            .map_err(|err| vm.new_value_error(format!("{}", err)))?;
```
Comment (coolreader18) [inline vm/src/obj/objstr.rs]:
```suggestion
            .map_err(|err| vm.new_value_error(err.to_string()))?;
```

Comment (ChJR) [reply vm/src/obj/objstr.rs]: That code found on str.__mod__ too. I suggest that code should be moved to another PR.

Comment (coolreader18) [reply vm/src/obj/objstr.rs]: Sounds good, thanks for contributing!

File: vm/src/obj/objstr.rs [hunk 4]
```diff
@@ -518,6 +518,18 @@ impl PyString {
...
+
+        match do_cformat(vm, _format_string, values.clone()) {
+            Ok(_format_string) => Ok(vm.ctx.not_implemented()),
```
Comment (windelbouwman) [inline vm/src/obj/objstr.rs]: Is the call to `do_cformat` required here? Or can we just always return `NotImplemented`?

Comment (ChJR) [reply vm/src/obj/objstr.rs]: I have just checked your second suggestion, and I found it works.
I used do_cformat just for argument checking, but that method is not required.
I will add a new commit for this PR.

**Comment by youknowone** [comment]
@ChJR there is a tiny style glitch. There is a clippy result in travis.

**Pull request merged**